extensions/python/PythonConfigState.h

```diff
@@ -0,0 +1,50 @@
+/**
+ *
+ * Licensed to the Apache Software Foundation (ASF) under one or more
+ * contributor license agreements.  See the NOTICE file distributed with
+ * this work for additional information regarding copyright ownership.
+ * The ASF licenses this file to You under the Apache License, Version 2.0
+ * (the "License"); you may not use this file except in compliance with
+ * the License.  You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+#pragma once
+
+#include <filesystem>
+#include <string>
+
+namespace org::apache::nifi::minifi::extensions::python {
+
+struct PythonConfigState {
+ public:
+  PythonConfigState(PythonConfigState&&) = delete;
+  PythonConfigState(const PythonConfigState&) = delete;
+  PythonConfigState& operator=(PythonConfigState&&) = delete;
+  PythonConfigState& operator=(const PythonConfigState&) = delete;
+
+  bool isPackageInstallationNeeded() const {
+    return install_python_packages_automatically && !virtualenv_path.empty();
+  }
+
+  static PythonConfigState& getInstance() {
+    static PythonConfigState config;
+    return config;
+  }
```

Comment:
I'd prefer no new singletons. Would it be possible? Is there any reason for this being a singleton?

Reply:
You are right, I also try to avoid them when possible. In this case the python configuration is read in the PythonCreator from the Configuration object, but has to be used in the PythonScriptEngine which is initialized in each python processor. Due to the interface we use to create the python processors (instantiating the ExecutePythonProcessor class from the PythonObjectFactory) there is no clean way currently to reach the `Configuration` object from the processors' scope. Even though it's better to avoid them I think this is still a cleaner solution than anything else I could come up with.

Reply:
Since PythonScriptEngine is created on a per-processor basis, we can't let it handle the creation of the shared execution environment. That should be extracted to a separate class, which could be owned by e.g. PythonCreator, setting up the virtualenv and packages, so that individual PythonScriptEngines can be created in that environment. PythonConfigState could be part of this new class, and PythonCreator could inject the Configuration object into it. What do you think about this?

Reply:
It makes sense in the context of this PR, I'm not sure what was the original intention of the `PythonScriptEngine::initialize` static function, as it was previously called with and empty bode from the `PythonCreator`, so my thought was that it was supposed to implement the global python environment specific initialization, but it can be moved to the `PythonCreator` or a separate class.

The problem is that in the PR depending on this https://github.com/apache/nifi-minifi-cpp/pull/1727 we introduce the support for inline python dependency installation, which is separately done for every NiFi python processor initialization that is called from the `ExecutePythonProcessor`. We also only want to initialize the processors that are part of the minifi flow config in the future. The inline python dependency installation needs the information stored in the `PythonConfigState` that is why it needs to be reached from outside the `PythonCreator`

Reply:
I think it could be solved by returning or moving the list of dependencies from the python processor to the execution environment class, which can install them as needed. A callback could probably work.

Reply:
Extracted dependency installation to the `PythonDependencyInstaller` class and moved the dependency install to `PythonCreator`, to install dependencies while processing all the python processor script files in 5b20de2ffdc4d8f2bfd8907d1a47294440a11fc2